temporal-sdk/src/main/java/io/temporal/internal/worker/ShutdownManager.java

```diff
@@ -0,0 +1,161 @@
+    return future;
+  }
+
+  public CompletableFuture<Void> shutdownExecutor(
+      ExecutorService executorToShutdown, String executorName) {
+    executorToShutdown.shutdown();
+
+    CompletableFuture<Void> future = new CompletableFuture<>();
+    scheduledExecutorService.scheduleWithFixedDelay(
+        new ReportingDelayShutdown(executorToShutdown, executorName, future),
+        0,
+        500,
+        TimeUnit.MILLISECONDS);
+    return future;
+  }
+
+  @Override
+  public void close() {
+    scheduledExecutorService.shutdownNow();
+  }
+
+  private static class LimitedWaitShutdown implements Runnable {
+    private final ExecutorService executorToShutdown;
+    private final CompletableFuture<Void> promise;
+    private final int maxAttempts;
+    private final String executorName;
+    private int attempt;
+
+    public LimitedWaitShutdown(
+        ExecutorService executorToShutdown,
+        int maxAttempts,
+        String executorName,
+        CompletableFuture<Void> promise) {
+      this.executorToShutdown = executorToShutdown;
+      this.promise = promise;
+      this.maxAttempts = maxAttempts;
+      this.executorName = executorName;
+    }
+
+    @Override
+    public void run() {
+      if (executorToShutdown.isTerminated()) {
+        promise.complete(null);
+        throw new RuntimeException("completed");
+      }
+      attempt++;
+      if (attempt > maxAttempts) {
+        log.warn(
+            "Wait for a graceful shutdown of {} timed out, fallback to shutdownNow", executorName);
+        executorToShutdown.shutdownNow();
+        // we don't want to complicate shutdown with dealing of exceptions and errors of all sorts,
+        // so just log and complete the promise
+        promise.complete(null);
+        throw new RuntimeException("timedout");
+      }
+    }
+  }
+
+  private static class ReportingDelayShutdown implements Runnable {
+    private static final int REPORTING_THRESHOLD = 30;
+    private static final int REPORTING_PERIOD = 10;
+
+    private final ExecutorService executorToShutdown;
+    private final CompletableFuture<Void> promise;
+    private final String executorName;
+    private int attempt;
+
+    public ReportingDelayShutdown(
+        ExecutorService executorToShutdown, String executorName, CompletableFuture<Void> promise) {
+      this.executorToShutdown = executorToShutdown;
+      this.promise = promise;
+      this.executorName = executorName;
+    }
+
+    @Override
+    public void run() {
+      if (executorToShutdown.isTerminated()) {
+        if (attempt > REPORTING_THRESHOLD) {
+          // log warn only if we already logged a shutdown being delayed
+          log.warn("{} successfully terminated", executorName);
+        }
+        promise.complete(null);
+        throw new RuntimeException("completed");
+      }
+      attempt++;
+      if (attempt >= REPORTING_THRESHOLD) {
+        if (((float) (attempt - REPORTING_THRESHOLD) / REPORTING_PERIOD) < 0.001) {
```

Comment:
Not exactly. You caught a bug here.
My intention was: After REPORTING_THRESHOLD I report one log each REPORTING_PERIOD to don't pollute the log too much. But it should be "%" instead of "/".